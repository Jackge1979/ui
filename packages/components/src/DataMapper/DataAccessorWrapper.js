import * as Constants from './Constants';

export default class DataAccessorWrapper {

  constructor(dataAccessor) {
    this.dataAccessor = dataAccessor;
  }

  /**
	* Returns the name of the schema.
	*/
	getSchemaName(schema) {
		return this.dataAccessor.getSchemaName(schema);
	}

	/**
	* Returns the number of elements in the schema.
	*/
	getSchemaSize(schema) {
		return this.dataAccessor.getSchemaSize(schema);
	}

	/**
	* Returns all the elements of the schema in an array.
	*/
	getSchemaElements(schema) {
		return this.dataAccessor.getSchemaElements(schema);
	}

	/**
	* Returns the nth element of the schema.
	*/
	getSchemaElement(schema, index) {
    if (this.dataAccessor.getSchemaElement) {
      return this.dataAccessor.getSchemaElement(schema, index);
    }
		const elements = this.getSchemaElements(schema);
    if (elements && elements.length > index) {
      return elements[index];
    }
    return null;
	}

	/**
	* Returns the index of the given element,
	* returns -1 if it is not in the schema.
	*/
	getSchemaElementIndex(schema, element) {
    if (this.dataAccessor.getSchemaElementIndex) {
      return this.dataAccessor.getSchemaElementIndex(schema, element);
    }
    const elements = this.getSchemaElements(schema);
    if (elements) {
      return elements.indexOf(element);
    }
		return -1;
	}

  getSchemaElementFromId(schema, id) {
    if (this.dataAccessor.getSchemaElementFromId) {
      return this.dataAccessor.getSchemaElementFromId(schema, id);
    }
    const elements = this.getSchemaElements(schema);
    if (elements) {
      return elements.find(elem => this.getElementId(elem) === id);
    }
    return null;
  }

	/**
	* Returns the identifier of the element.
	* Identifier must be unique.
	*/
	getElementId(element) {
		return this.dataAccessor.getElementId(element);
	}

	/**
	* Returns true if the two elements are equals.
	* Default implementation is based on element id.
	*/
	areEquals(element1, element2) {
    if (this.dataAccessor.areEquals) {
      return this.dataAccessor.areEquals(element1, element2);
    }
		return this.getElementId(element1) === this.getElementId(element2);
	}

	/**
	* Returns true if the array of elements contains the specified element.
	*/
	includes(elements, element) {
    if (this.dataAccessor.includes) {
      return this.dataAccessor.includes(elements, element);
    }
    for (let i = 0; i < elements.length; i += 1) {
      if (this.areEquals(elements[i], element)) {
        return true;
      }
    }
    return false;
	}

	/**
	* Returns the name of the element.
	*/
	getElementName(element) {
		return this.dataAccessor.getElementName(element);
	}

  haveSameName(element1, element2) {
  	return this.getElementName(element1) === this.getElementName(element2);
  }

	/**
	* Returns the semantic type of the element.
	*/
	getElementType(element) {
		return this.dataAccessor.getElementType(element);
	}

	/**
	* Returns a description fot the element.
	*/
	getElementDescription(element) {
		return this.dataAccessor.getElementDescription(element);
	}

	/**
	* Returns an array of all the mapping items. A mapping item defines
	* a mapping betwwen an input and an output element.
	*/
	getMappingItems(mapping) {
		return this.dataAccessor.getMappingItems(mapping);
	}

  /**
  * Returns true if the given mapping is empty
  */
  isMappingEmpty(mapping) {
    if (this.dataAccessor.isMappingEmpty) {
      return this.dataAccessor.isMappingEmpty(mapping);
    }
    const items = this.getMappingItems(mapping);
    return !items || !items.length;
  }

	/**
	* Return the mapped element for the specified side.
	*/
	getMappedElement(mappingItem, side) {
		return this.dataAccessor.getMappedElement(mappingItem, side);
	}

	/**
	* Add a new mapping from source element to target element.
	* Returns the updated mapping.
	*/
	addMapping(mapping, source, target) {
		return this.dataAccessor.addMapping(mapping, source, target);
	}

	/**
	* Remove the (source->target) mapping.
	* Returns the updated mapping.
	*/
	removeMapping(mapping, source, target) {
		return this.dataAccessor.removeMapping(mapping, source, target);
	}

	/**
	* Remove all the mapping items.
	* Returns the updated mapping.
	*/
	clearMapping(mapping) {
		return this.dataAccessor.clearMapping(mapping);
	}

  // Some more convenient methods

  isElementInMappingItem(mappingItem, element, side) {
  	const mappedElement = this.getMappedElement(mappingItem, side);
  	return this.areEquals(mappedElement, element);
  }

  /**
   * isElementMapped returns true if the given (element, side) is mapped
   * (i.e. if it appears in the mapping)
   */
  isElementMapped(mapping, element, side) {
  	if (mapping != null) {
  		const mappingItems = this.getMappingItems(mapping);
  		return mappingItems.find(
  			item => this.isElementInMappingItem(item, element, side)
  		);
  	}
  	return false;
  }

  /**
  * isFullMapped returns true if all the elements of the given schema are mapped
  */
  isFullMapped(mapping, schema, side) {
  	// TODO could be optimized
  	for (let i = 0; i < this.getSchemaSize(schema); i += 1) {
  		if (!this.isElementMapped(mapping, this.getSchemaElement(schema, i), side)) {
  			return false;
  		}
  	}
  	return true;
  }

  getMappingItemsWithElement(mapping, element, side) {
  	const mappingItems = this.getMappingItems(mapping);
  	return mappingItems.filter(item => this.isElementInMappingItem(item, element, side));
  }

  getConnectedElements(mapping, element, side) {
  	const items = this.getMappingItemsWithElement(mapping, element, side);
  	if (items != null) {
  		const connectedSide = Constants.switchMappingSide(side);
  		return items.map(item => this.getMappedElement(item, connectedSide));
  	}
  	return null;
  }

}
