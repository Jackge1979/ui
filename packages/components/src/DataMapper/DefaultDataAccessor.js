import * as Constants from './Constants';

/**
 * Internal function only used by the dataAccessor.
 */
function getMappingItemIndex(mapping, source, target) {
	return mapping.findIndex(item =>
		item.source.id === source.id && item.target.id === target.id);
}

/**
 * Internal function only used by the dataAccessor.
 */
function removeMappingItem(mapping, index) {
	const updatedMapping = mapping.slice();
	updatedMapping.splice(index, 1);
	return updatedMapping;
}

/**
 * This interface provides all the needed methods to access/update the
 * schema and mapping data.
 */
export default class DefaultDataAccessor {
	/**
	 * Returns the name of the schema.
	 */
	getSchemaName(schema) {
		return schema.name;
	}

	/**
	 * Returns the number of elements in the schema.
	 */
	getSchemaSize(schema) {
		return schema.elements.length;
	}

	/**
	 * Returns all the elements of the schema in an array.
	 */
	getSchemaElements(schema) {
		return schema.elements;
	}

	/**
	 * Returns the nth element of the schema.
	 */
	getSchemaElement(schema, index) {
		return schema.elements[index];
	}

	/**
	 * Returns the index of the given element,
	 * returns -1 if it is not in the schema.
	 */
	getSchemaElementIndex(schema, element) {
		return schema.elements.indexOf(element);
	}

	/**
	 * Returns the identifier of the element.
	 * Identifier must be unique.
	 */
	getElementId(element) {
		return element.id;
	}

	/**
	 * Returns the name of the element.
	 */
	getElementName(element) {
		return element.name;
	}

	/**
	 * Returns the semantic type of the element.
	 */
	getElementType(element) {
		return element.type;
	}

	/**
	 * Returns a description fot the element.
	 */
	getElementDescription(element) {
		return element.description;
	}

	/**
	 * Returns an array of all the mapping items. A mapping item defines
	 * a mapping betwwen an input and an output element.
	 */
	getMappingItems(mapping) {
		return mapping;
	}

	/**
	 * Return the mapped element for the specified side.
	 */
	getMappedElement(mappingItem, side) {
		if (side === Constants.MappingSide.INPUT) {
			return mappingItem.source;
		}
		return mappingItem.target;
	}

	/**
	 * Add a new mapping from source element to target element.
	 * Returns the updated mapping.
	 */
	addMapping(mapping, source, target) {
		return mapping.concat({ source, target });
	}

	/**
	 * Remove the (source->target) mapping.
	 * Returns the updated mapping.
	 */
	removeMapping(mapping, source, target) {
		const index = getMappingItemIndex(mapping, source, target);
		if (index >= 0) {
			return removeMappingItem(mapping, index);
		}
		return mapping;
	}

	/**
	 * Remove all the mapping items.
	 * Returns the updated mapping.
	 */
	clearMapping(mapping) {
		return [];
	}

	/**
	 * Returns true if the given mapping is empty
	 */
	isMappingEmpty(mapping) {
		return !mapping.length;
	}

	/**
	 * Returns true if the array of elements contains the specified element.
	 */
	includes(elements, element) {
		return elements.find(elem => elem.id === element.id);
	}
}
