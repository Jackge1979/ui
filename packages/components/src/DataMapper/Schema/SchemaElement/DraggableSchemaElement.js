import React, { Component } from 'react';
import { ItemTypes, SchemaType } from '../../Constants';
import { DragSource } from 'react-dnd';
import { DropTarget } from 'react-dnd';
import PropTypes from 'prop-types';
import flow from 'lodash/flow';
import SchemaElement from './SchemaElement.js';

const elementSource = {
	beginDrag(props) {
		return {
			elementId: props.name,
			schemaType: props.schemaType,
		};
	},
};

const elementTarget = {
	drop(props, monitor) {
		var sourceElem = monitor.getItem();
		props.performMapping(sourceElem.elementId, props.name);
	},
	canDrop(props, monitor) {
		var sourceElem = monitor.getItem();
		return props.schemaType !== sourceElem.schemaType && sourceElem.schemaType === SchemaType.INPUT;
	},
};

function collectForDragSource(connect, monitor) {
	return {
		connectDragSource: connect.dragSource(),
		isDragging: monitor.isDragging(),
	};
}

function collectForDropTarget(connect, monitor) {
	return {
		connectDropTarget: connect.dropTarget(),
		dragOver: monitor.isOver(),
	};
}

class DraggableSchemaElement extends Component {
	render() {
		const {
			name,
			schemaType,
			mapped,
			connectDragSource,
			connectDropTarget,
			onSelect,
			selected,
		} = this.props;
		const highlighted = this.props.dragOver;
		return connectDragSource(
			connectDropTarget(
				<div>
					<SchemaElement
						highlighted={highlighted}
						mapped={mapped}
						name={name}
						schemaType={schemaType}
						selected={selected}
						onSelect={onSelect}
					/>
				</div>,
			),
		);
	}
}

DraggableSchemaElement.propTypes = {
	name: PropTypes.string,
	schemaType: PropTypes.string,
};

export default flow(
	DragSource(ItemTypes.ELEMENT, elementSource, collectForDragSource),
	DropTarget(ItemTypes.ELEMENT, elementTarget, collectForDropTarget),
)(DraggableSchemaElement);
