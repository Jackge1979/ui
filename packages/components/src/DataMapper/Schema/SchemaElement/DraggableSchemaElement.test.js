import React from 'react';
import renderer from 'react-test-renderer';
import DraggableSchemaElement from './DraggableSchemaElement.js';
import DefaultDataAccessor from '../../DefaultDataAccessor';
import DataAccessorWrapper from '../../DataAccessorWrapper';

const dataAccessor = new DataAccessorWrapper(new DefaultDataAccessor());

it('single-element', () => {
	// Obtain the reference to the component before React DnD wrapping
	const OriginalSchemaElement = DraggableSchemaElement.DecoratedComponent;
	// Stub the React DnD connector functions with an identity function
	const identity = el => el;
	// create React tree
	const tree = renderer
		.create(
			<OriginalSchemaElement
				dataAccessor={dataAccessor}
				element="Single_element"
				connectDragSource={identity}
				connectDropTarget={identity}
			/>,
		)
		.toJSON();
	expect(tree).toMatchSnapshot();
});

it('mapped-element', () => {
	// Obtain the reference to the component before React DnD wrapping
	const OriginalSchemaElement = DraggableSchemaElement.DecoratedComponent;
	// Stub the React DnD connector functions with an identity function
	const identity = el => el;
	// create React tree
	const tree = renderer
		.create(
			<OriginalSchemaElement
				dataAccessor={dataAccessor}
				element="Mapped_element"
				mapped={true}
				connectDragSource={identity}
				connectDropTarget={identity}
			/>,
		)
		.toJSON();
	expect(tree).toMatchSnapshot();
});

it('drag-over-element', () => {
	// Obtain the reference to the component before React DnD wrapping
	const OriginalSchemaElement = DraggableSchemaElement.DecoratedComponent;
	// Stub the React DnD connector functions with an identity function
	const identity = el => el;
	// create React tree
	const tree = renderer
		.create(
			<OriginalSchemaElement
				dataAccessor={dataAccessor}
				element="Drag_over_element"
				dragOver={true}
				connectDragSource={identity}
				connectDropTarget={identity}
			/>,
		)
		.toJSON();
	expect(tree).toMatchSnapshot();
});
