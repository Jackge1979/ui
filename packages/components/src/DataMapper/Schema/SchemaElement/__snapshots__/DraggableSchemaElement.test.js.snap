// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`drag-over-element 1`] = `
<div>
  <div
    className="schema-element highlighted undefined"
    onClick={[Function]}
    role="button"
  >
    Drag_over_element
  </div>
</div>
`;

exports[`mapped-element 1`] = `
<div>
  <div
    className="schema-element mapped undefined"
    onClick={[Function]}
    role="button"
  >
    Mapped_element
  </div>
</div>
`;

exports[`single-element 1`] = `
<div>
  <div
    className="schema-element undefined"
    onClick={[Function]}
    role="button"
  >
    Single_element
  </div>
</div>
`;
